docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+
+* Design should take into account the Cryptographic Token Interface Standard (PKCS#11)
+
+Increase cryptographic versatility
+
+* Support for a broad spectrum of cryptographic techniques
+* Extend support for more hash functions (e.g. pedersen, argon2, Argon2d/I/id, Blake3, etc.)
+* Extend support for more signature schemes (e.g. secp256r1, ed25519, ed448, sr25519, etc.)
+* More advanced methods ( Post-Quantum Cryptography (PQC) methods
+* Threshold signatures and encryption
+
+Community Engagement Infrastructure:
+
+* Structure the design with tools and documentation interfaces in mind, enabling a seamless future rollout of resources for developer engagement.
+
+## Proposed architecture
+
+### Introduction
+
+In the proposed architecture, each package is decoupled and isolated. Adding new implementations consist of implementing the required interfaces.
+
+```mermaid
+classDiagram
+
+Keyring <|-- Wallet
+
+SecureStorage <|-- Keyring
+SecureItem <|-- SecureStorage
+CryptoProvider <|-- SecureItem
+
+Hasher <|-- CryptoProvider
+Verifier <|-- CryptoProvider
+Signer <|-- CryptoProvider
+Cipher <|-- CryptoProvider
+Generator <|-- CryptoProvider
+```
+
+### Crypto Provider
+
+*Crypto Providers* serve as a middleware responsible for managing the interaction with various instantiated cryptographic packages. It acts as a centralized controller, encapsulating the API of the crypto modules in a single location.
+Through each Crypto provider, users can access functionality such as signing, verification, encryption, and hashing.
+
+By abstracting the underlying cryptographic functionality, *Crypto providers* enable a modular and extensible architecture. It allows users to easily switch between different cryptographic implementations without impacting the rest of the system.
+
+A `CryptoProvider` can be built from different sources, such as a `SecureItem`, a randomness source, a seed, a mnemonic, or a string.
+Since the underlying representation of a `CryptoProvider` is a protobuf message, it can be easily serialized and stored in a safe manner.
+The storing and retrieval is handled by a `SecureStorage` implementation which will be presented in the next section.
+
+```go
+type ProviderMetadata interface {
+  key string
+  value string
+}
+
+type ICryptoProviderMetadata interface {
+  GetTypeUUID() TypeUUID
+  GetName() string
+  GetMetadata() []ProviderMetadata
+}
+
+type ICryptoProviderBuilder interface {
+  ICryptoProviderMetadata
+  
+  FromSecureItem( item SecureItem ) (ICryptoProvider, error)  
+
+  FromRandomness( source IRandomnessSource ) (ICryptoProvider, error)
+  FromSeed( seed []byte ) (ICryptoProvider, error)
+  FromMnemonic( mnemonic string ) (ICryptoProvider error)
+  FromString( url string ) (ICryptoProvider error)
+}
+
+type ICryptoProvider interface {
+  Proto.Message
+  ICryptoProviderMetadata
+  
+  GetKeys() (PubKey, PrivKey, error)
```

Comment:
If I understand correctly, an ICryptoProvider implementation is backed by a specific keypair, correct?

What do its Cipher and Hasher represent then? What algorithm do they implement?

Reply:
yes, each crypto provider is associated to a keypair in the SecureItem struct, like described in https://github.com/cosmos/cosmos-sdk/pull/18824#discussion_r1562781586